processor/src/networks/bitcoin.rs

```diff
@@ -450,6 +477,85 @@ impl Bitcoin {
       }
     }
   }
+
+  // expected script has to start with SHA256 PUSH MSG_HASH OP_EQ ..
+  fn expected_script_pattern(&self, script: &ScriptBuf) -> Option<bool> {
+    let mut ins = script.instructions();
+
+    // first item should be SHA256 code
+    if ins.next()?.ok()?.opcode()? != OP_SHA256 {
+      return Some(false);
+    }
+
+    // next should be a data push
+    ins.next()?.ok()?.push_bytes()?;
+
+    // next should be a equality check
+    if ins.next()?.ok()?.opcode()? != OP_EQUAL {
+      return Some(false);
+    }
+
+    Some(true)
+  }
+
+  async fn spent_output_of(&self, input: &TxIn) -> TxOut {
+    let mut spent_tx = input.previous_output.txid.as_raw_hash().to_byte_array();
+    spent_tx.reverse();
+    let mut tx;
+    while {
+      tx = self.get_transaction(&spent_tx).await;
+      tx.is_err()
+    } {
+      log::error!("couldn't get transaction from bitcoin node: {tx:?}");
+      sleep(Duration::from_secs(5)).await;
+    }
+    tx.unwrap().output.swap_remove(usize::try_from(input.previous_output.vout).unwrap())
+  }
+
+  async fn extract_serai_data(&self, tx: &Transaction) -> Vec<u8> {
+    // check outputs
+    let mut data = (|| {
+      for output in &tx.output {
+        if output.script_pubkey.is_op_return() {
+          match output.script_pubkey.instructions_minimal().last() {
+            Some(Ok(Instruction::PushBytes(data))) => return data.as_bytes().to_vec(),
+            _ => continue,
+          }
+        }
+      }
+      vec![]
+    })();
+
+    // check inputs
+    if data.is_empty() {
+      for input in &tx.input {
+        if self.spent_output_of(input).await.script_pubkey.is_p2wsh() {
```

Comment:
I don't believe this is a necessary check. Checking alone the witness is as expected *should* be sufficient. If this isn't *some* SegWit TX, then witness will be empty.

It does mean we may interpret a non-P2WSH witness as a P2WSH witness. I'm unsure that's an issue. The person who sent this TX still gave us something we can interpret *and they included it and therefore authorized it*.

Considering how expensive the spent_output check is (it's pulling an old TX which is a full RPC call and an extra user of the txindex functionality of the underlying node), I don't see the benefit unless there's an explicit issue at risk.

Reply:
To be clear, we use txindex which adds GB of data to the Bitcoin node because it's currently impractical to track everything on our end. Ideally, yes, we don't need txindex to minimize data usage and improve performance.

If usage of it makes sense, I don't mind using it. If usage of it is extraneous (pushing us further away from being able to remove it), I can't claim to like it.